### ISSUE TITLE
Lock goes crazy with dependencies specified multiple times.

### ISSUE THREAD

**Issue by SirMartin** [pdm-project/pdm#1311]
- [X] I have searched the issue tracker and believe that this is not a duplicate.

**Make sure you run commands with `-v` flag before pasting the output.**

## Steps to reproduce

Not sure if this is a bug or a config error, I saw the same behaviour in pip, but I think not in pipenv, in the same project.

I have a project with these dependencies in the pyproject.toml

```
    "ddtrace==0.49.3",
    "ddtrace[opentracing]",
    "opentracing==2.4.0",
```

With this configuration, it takes more than 7 minutes to make the lock, because PDM tries to find in all the options (if I make pip install -r requirements.txt) it makes the same. I pasted down there, a bit of the example of the step, but there are hundreds of lines with the same message per each version check (almost 7000 lines).

```
pdm.termui: ======== Starting round 22 ========
pdm.termui:   Adding requirement opentracing>=2.0.0(from ddtrace 1.3.2)
pdm.termui:   Adding requirement ddtrace==1.3.2(from ddtrace 1.3.2)
pdm.termui:   Adding requirement opentracing>=2.0.0(from ddtrace 1.3.2)
pdm.termui:   Adding requirement ddtrace==1.3.2(from ddtrace 1.3.2)
pdm.termui:   Adding requirement opentracing>=2.0.0(from ddtrace 1.3.2)
pdm.termui:   Adding requirement ddtrace==1.3.2(from ddtrace 1.3.2)
pdm.termui:   Adding requirement opentracing>=2.0.0(from ddtrace 1.3.2)
```

Check the attached log file to get the full trace: [pdm-attempts-ddtrace.log](https://github.com/pdm-project/pdm/files/9288992/pdm-attempts-ddtrace.log)

I found trying to convert this project to PDM, that if I remove `ddtrace[opentracing]` or `opentracing==2.4.0` from the locking, it take just around 15 seconds to make it, and don´t make any other attempts to find versions. 

In all the cases install opentracing in the version 2.4.0, even when is not specified (2.4.0 is the latest available version of the library).

It can be useful too, that I am using also `basictracer==3.1.0` that has `opentracing<3.0,>=2.0` as requirement.

## Actual behavior

PDM tries hundreds of attempts in the lock process to find which installation of ddtrace and opentracing make.

## Expected behavior

I think it should not take forever, or maybe stop it after some time, if cannot find a version.

## Environment Information

```bash
# Paste the output of `pdm info && pdm info --env` below:
PDM version:
  2.1.1
Python Interpreter:

  /usr/bin/python3.9 (3.9)
Project Root:
  /home/edu/NextailGit/py-common
Project Packages:
  /home/edu/NextailGit/py-common/__pypackages__/3.9

{
  "implementation_name": "cpython",
  "implementation_version": "3.9.13",
  "os_name": "posix",
  "platform_machine": "x86_64",
  "platform_release": "5.17.15-76051715-generic",
  "platform_system": "Linux",
  "platform_version": "#202206141358~1655919116~22.04~1db9e34 SMP PREEMPT Wed 
Jun 22 19",
  "python_full_version": "3.9.13",
  "platform_python_implementation": "CPython",
  "python_version": "3.9",
  "sys_platform": "linux"
}


```

**Comment by frostming** [comment]
This should be resolved in the user side, closing it now.

**Comment by SirMartin** [comment]
> Wouldn't `"ddtrace[opentracing]==0.49.3"` suffice?

Probably yes, or just specify the other version, and fix the issue. I just reported because I was not sure was the expected behaviour from PDM side.

But I really don´t understand when ddtrace==0.49.3 is specified, tries all the other ddtrace versions, when it should be ok take just the version 0.49.3

Thanks for the help

**Comment by frostming** [comment]
> But I really don´t understand when ddtrace==0.49.3 is specified, tries all the other ddtrace versions, when it should be ok take just the version 0.49.3

Because the resolver takes `ddtrace[opentracing]` as a different requirement and hence doesn't reuse the resolved version of `0.49.3`, it thinks it is a new requirement **without version constraints**. What is unique about it is it produces a virtualenv dependency on `ddtrace[opentracing] -> ddtrace`, so eventually only one version of `ddtrace` will get pinned. But before that, all versions of it must be looked at.

I know the resolution approach is sub-optimal but this is what the situation is for now.

**Comment by SirMartin** [comment]
Ok, understood. Good to know.
